Add tutorials for blend animation and retargeting

TokageItLab opened the pull request:
Added a section about the Blend animation which has been changed significantly in Godot 4.0, and a new page describing about the importer retarget.

In particular, the scene importer does not have a class reference, so it is necessary to explain it in the documentation.

Fixes https://github.com/godotengine/godot/issues/59536
Fixes https://github.com/godotengine/godot/issues/65771

fire commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,38 @@ Transitions also have a few properties. Click any transition and it will be disp
+For better blending
+-------------------
+
+In Godot 4.0+, for blending results consistency, the property values to be blended always have a specific initial value. For example, in the case of two animations to be blended, if one animation has a property track and the other does not, the blended animation is calculated as if the latter animation had a property track with the initial value.
```
for blending results consistency, -> for blending results consistently,

fire replied:
```suggestion
In Godot 4.0+, for blending results consistently, the property values to be blended always have a specific initial value. For example, in the case of two animations to be blended, if one animation has a property track and the other does not, the blended animation is calculated as if the latter animation had a property track with the initial value.
```

fire commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,38 @@ Transitions also have a few properties. Click any transition and it will be disp
+
+In Godot 4.0+, for blending results consistency, the property values to be blended always have a specific initial value. For example, in the case of two animations to be blended, if one animation has a property track and the other does not, the blended animation is calculated as if the latter animation had a property track with the initial value.
+
+For Position/Rotation/Scale 3D tracks for Skeleton3D bones, the initial value is Bone Rest. For other properties, the initial value is ``0`` (exactly the value of ``zero()`` as Variant) and if present the track value in RESET animation, it's first key value is used instead.
```
```suggestion
When using Position/Rotation/Scale 3D tracks for Skeleton3D bones, the initial value is Bone Rest. For other properties, the initial value is ``0`` (exactly the value of ``zero()`` as Variant) and if present the track value in RESET animation, it's first key value is used instead.
```

fire commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,38 @@ Transitions also have a few properties. Click any transition and it will be disp
+
+.. image:: img/blending5.gif
+
+You can see that the shortest rotation path from Bone Rests is taken priority rather than the shortest rotation path between animations.
```
```suggestion
You can see that the shortest rotation path from Bone Rests is prioritized rather than the shortest rotation path between animations.
```

fire commented on tutorials/assets_pipeline/index.rst:
```diff
@@ -10,5 +10,6 @@ Assets pipeline
    importing_audio_samples
    importing_translations
    importing_scenes
+   retargeting_3d_skeleton
```
```suggestion
   retargeting_3d_skeletons
```

fire commented on tutorials/assets_pipeline/retargeting_3d_skeleton.rst:
```diff
@@ -0,0 +1,143 @@
+.. _doc_retargeting_3d_skeleton:
```
```suggestion
.. _doc_retargeting_3d_skeletons:
```

lyuma commented on tutorials/assets_pipeline/retargeting_3d_skeleton.rst:
```diff
@@ -0,0 +1,143 @@
+
+.. image:: img/retargeting1.png
+
+Therefore, first set up a new BoneMap and SkeletonProfile. Godot has a preset SkeletonProfileHumanoid for humanoid models. This tutorial proceed
```
```suggestion
With the Skeleton node selected in the importer, first set up a new BoneMap and SkeletonProfile. Godot has a preset SkeletonProfileHumanoid for humanoid models. This tutorial will proceed
```

lyuma commented on tutorials/assets_pipeline/retargeting_3d_skeleton.rst:
```diff
@@ -0,0 +1,143 @@
+Fix Silhouette
+^^^^^^^^^^^^^^
+
+Makes the model silhouette closer to the Reference Pose silhouette defined in the SkeletonProfile. This may not work well if the silhouettes are too different or for fixing roll.
```
```suggestion
Attempts to make the model's silhouette match that of the reference pose defined in the SkeletonProfile, such as T-Pose. This cannot fix silhouettes which are too different, and it may not work for fixing bone roll.
```

lyuma commented on tutorials/assets_pipeline/retargeting_3d_skeleton.rst:
```diff
@@ -0,0 +1,143 @@
+
+Makes the model silhouette closer to the Reference Pose silhouette defined in the SkeletonProfile. This may not work well if the silhouettes are too different or for fixing roll.
+
+With SkeletonProfileHumanoid, this option does not need to be enabled for T-pose models, but should be enabled for A-pose models. However in that case, the fixed foot results may be bad depends on the heel height of the model, so it may be necessary to specify the bone names of the SkeletonProfile in the Filter for which you do not want to fix.
```
```suggestion
With SkeletonProfileHumanoid, this option does not need to be enabled for T-pose models, but should be enabled for A-pose models. However in that case, the fixed foot results may be bad depending on the heel height of the model, so it may be necessary to add the SkeletonProfile bone names you do not want fixed in the Filter array, as in the below example.
```

mhilbrunner commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,47 @@ Transitions also have a few properties. Click any transition and it will be disp
+For better blending
+-------------------
+
+In Godot 4.0+, for blending results consistently, the property values to be blended always have a specific initial value.
```
```suggestion
In Godot 4.0+, for blending results consistently, the property values to be blended always have a specific initial value.
```
Grammar feels funky here, but I'm not enough of a natural speaker to provide a suggestion.
Maybe "for blending results to be consistent", or maybe restructure the whole sentence. Not sure. Hope someone else has a suggestion :)

TokageItLab replied:
@mhilbrunner In Godot 3, when blending was performed during animation playback, the blending result depended on when the blending was performed.

I want to mean that the results of blending always be equal.

mhilbrunner commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,47 @@ Transitions also have a few properties. Click any transition and it will be disp
+the blended animation is calculated as if the latter animation had a property track with the initial value.
+
+When using Position/Rotation/Scale 3D tracks for Skeleton3D bones, the initial value is Bone Rest.
+For other properties, the initial value is ``0`` (exactly the value of ``zero()`` as Variant)
```
```suggestion
(exactly the value of ``zero()`` as Variant)
```
Is this a neccessary detail? It IMO makes an already complicated topic/sentence more complex.

TokageItLab replied:
@mhilbrunner Would you recommend that I remove the `zero()` statement?

It was meant to suggest that there are two initial values for Variant, init() and zero(), but that we need to use zero() for correct calculation, but it is okay to remove it.

Position/Rotation/Scale 3D tracks for bones use Bone Rest. In this case, Bone Rest is implicitly and fundamentally never set to 0 on import.

However, Property tracks for changing other properties may become 0 unless RESET animation is explicitly set.

TokageItLab replied:
Removed mention about zero().

mhilbrunner commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,47 @@ Transitions also have a few properties. Click any transition and it will be disp
+          It is assumed to have only one frame and is not expected to be played back using the timeline.
+
+Also keep in mind that the Rotation 3D tracks and the Property tracks for 2D rotation
+with Interpolation Type set to Linear Angle or Cubic Angle will prevent rotation more than 180 degrees
```
```suggestion
with Interpolation Type set to Linear Angle or Cubic Angle will prevent rotation of more than 180 degrees
```

mhilbrunner commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,47 @@ Transitions also have a few properties. Click any transition and it will be disp
+with Interpolation Type set to Linear Angle or Cubic Angle will prevent rotation more than 180 degrees
+from the initial value as blended animation.
+
+This behavior is useful for Skeleton3D to prevent the bones penetrating the body when blend animations.
```
```suggestion
This can be useful for Skeleton3Ds to prevent the bones penetrating the body when blending animations.
```

mhilbrunner commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,47 @@ Transitions also have a few properties. Click any transition and it will be disp
+
+This behavior is useful for Skeleton3D to prevent the bones penetrating the body when blend animations.
+Therefore, Skeleton3D's Bone Rest values should be as close to the midpoint of the movable range as possible.
+**This means that for humanoid models, it is preferable to import them with T-pose**.
```
```suggestion
**This means that for humanoid models, it is preferable to import them in a T-pose**.
```

mhilbrunner commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,47 @@ Transitions also have a few properties. Click any transition and it will be disp
+Therefore, Skeleton3D's Bone Rest values should be as close to the midpoint of the movable range as possible.
+**This means that for humanoid models, it is preferable to import them with T-pose**.
+
+.. image:: img/blending5.gif
```
Before merging this, the docs team needs to make a decisions on the move to .webp, as an animated webp could potentially save a lot of size here compared to a 4-5mb gif.

TokageItLab replied:
I pushed webp to test it out.

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+To share animations among multiple Skeletons
+--------------------------------------------
+
+Godot has Position/Rotation/Scale 3D tracks (hereafter called Transform track)
```
```suggestion
Godot has Position/Rotation/Scale 3D tracks (hereafter called Transform tracks)
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+--------------------------------------------
+
+Godot has Position/Rotation/Scale 3D tracks (hereafter called Transform track)
+with Nodepath to bone for Skeleton bone animation. If you think that you can
```
```suggestion
with Nodepaths to bones for Skeleton bone animation. If you think that you can
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+same, that is not correct.
+
+Godot allows each bone to have a parent-child relationship and can have rotation
+and scale as well as position. It means that the bones can have Transform values
+independently.
```
```suggestion
and scale as well as position, which means that the bones can have independent Transform values.
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+
+.. note:: Godot 3.x and Godot 4.0+ have different Bone Pose behaviors.
+          In Godot 3.x, Bone Pose is relative to Bone Rest, but in Godot 4.0+,
+          it includes Bone Rest. See the `article <https://godotengine.org/article/animation-data-redesign-40>`__.
```
```suggestion
          it includes Bone Rest. See this `article <https://godotengine.org/article/animation-data-redesign-40>`__.
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+and scale as well as position. It means that the bones can have Transform values
+independently.
+
+Skeleton has the Transform values necessary for the default pose as Bone Rest.
```
```suggestion
The Skeleton holds the Transform values necessary for the default pose as Bone Rest.
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+models without any Bone Rest rotations, such as the glTF model output from Maya.
+
+To share animations in Godot, it is necessary to match Bone Rests as well as Bone Names,
+moreover to remove unwanted tracks in some cases. In Godot 4.0+, you can do that workflow
+on the scene importer.
```
```suggestion
moreover to remove unwanted tracks in some cases. In Godot 4.0+, you can do that using
the scene importer.
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+moreover to remove unwanted tracks in some cases. In Godot 4.0+, you can do that workflow
+on the scene importer.
+
+Options for Retarget
+--------------------
```
```suggestion
Options for Retargeting
-----------------------
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+.. image:: img/retargeting1.png
+
+With the Skeleton node selected in the importer, first set up a new BoneMap and SkeletonProfile.
+Godot has a preset SkeletonProfileHumanoid for humanoid models.
```
```suggestion
Godot has a preset called SkeletonProfileHumanoid for humanoid models.
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+
+.. image:: img/retargeting2.png
+
+Any missing / duplicate or incorrect parent-child relationship mappings will be indicated
```
```suggestion
Any missing, duplicate or incorrect parent-child relationship mappings will be indicated
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+.. note:: The auto-mapping uses pattern matching for the bone names. So we recommend
+          to use common English names for bones.
+
+After you set up the BoneMap, several options are available in the below sections.
```
```suggestion
After you set up the BoneMap, several options are available in the sections below.
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+Removes Position tracks other than ``root_bone`` and ``scale_base_bone``
+defined in SkeletonProfile from the animations. In SkeletonProfileHumanoid,
+this means that to remove Position tracks other than Root and Hips.
+Since Godot 4.0+ animations include Bone Rest in the transform value, if you disable this option,
```
```suggestion
Since Godot 4.0+, animations include Bone Rest in the transform value, if you disable this option,
```

mhilbrunner commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,171 @@
+* Directs the +Y axis from the parent joint to the child joint
+* +X rotation bends the joint like a muscle contracting
+
+These rules are convenient definitions for blend animation and IK.
```
```suggestion
These rules are convenient definitions for blend animation and Inverse Kinematics (IK).
```

mhilbrunner commented:
Left a few comments. All in all, I feel like this PR needs some attention by a natural speaker, as I kind of struggled with whether my difficulty in following along was because of the language or due to my lack of familiarity with the topic.
I've left notes where I could to indicate where I think improvements could be made, but if someone like @clayjohn or someone else could give this a glance too, I'd be thankful. :)

Besides that, we need to come to a decision regarding the use of .webp for images and animated images, I'll try to get that done this week before we merge this.

Thanks for contributing!

TokageItLab commented on tutorials/animation/animation_tree.rst:
```diff
@@ -188,6 +188,47 @@ Transitions also have a few properties. Click any transition and it will be disp
+For better blending
+-------------------
+
+In Godot 4.0+, for blending results consistently, the property values to be blended always have a specific initial value.
```
I don't think it is necessary to mentions even the internal process (the setting order and the connecting order) since it increases the complexity.

TokageItLab replied:
How about this?

> In Godot 4.0+ to make blending results reproducible and consistent always, the blended property values to be blended must have a specific initial value.

fire replied:
Try

The blended property values must have a specific initial value to make blending results reproducible and consistent in Godot.

TheYellowArchitect replied:
I may try building this tutorial within a few days, and trying it out to give feedback. Until then, I cannot resist not giving some feedback on that phrase

>In Godot 4.x, in order for the blending results to be deterministic (reproducible and always consistent), the blended property values must have a specific initial value

mhilbrunner reviewed (commented):
We just decided on the move to .webp, and seeing the file size savings here, this indeed seems to be a good improvement :)

If you can replace the .pngs with .webp (lossless if <250kb, otherwise use quality settings that seem like the best fit) and squash commits, that'd be appreciated :)

TokageItLab commented:
@mhilbrunner

> quality settings

Is there clear criteria like a target size?

TokageItLab commented:
@mhilbrunner I changed the images to webp and squashed commits. There is no clear target size given for the animated images, but I compressed them with the ffmpeg default of 75 for the time being. It appears a bit noisy, but I think it's acceptable. If necessary, I will re-encode with cleaner settings.

mhilbrunner commented:
Nice work! This is currently in the Asset Pipeline section, which makes sense as it mostly relates to import settings and importing skeletons I guess (from our/the engine's perspective).

Question is if a new user would be more likely to look under the Animation section? Opinions? I'm fine with leaving it where it is, just want to bring it up.

clayjohn commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,170 @@
+
+.. image:: img/retargeting1.webp
+
+With the Skeleton node selected in the importer, first set up a new BoneMap and SkeletonProfile.
+Godot has a preset called SkeletonProfileHumanoid for humanoid models.
+This tutorial will proceed with the assumption that you will use SkeletonProfileHumanoid.
```
```suggestion
With the Skeleton node selected, first set up a new BoneMap and SkeletonProfile.
Godot has a preset called SkeletonProfileHumanoid for humanoid models.
This tutorial proceeds with the assumption that you are using SkeletonProfileHumanoid.
```

clayjohn commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,170 @@
+.. note:: If you need a profile for a model such as a beast or creature, you can export
+          a SkeletonProfile from SkeletonEditor with some information converted from Skeleton.
+
+When you use SkeletonProfileHumanoid, the auto-mapping will be performed when the
+SkeletonProfile is set. If the auto-mapping does not work well, you can map bones manually.
```
```suggestion
When you use SkeletonProfileHumanoid, auto-mapping will be performed when the
SkeletonProfile is set. If the auto-mapping does not work well, you can map bones manually.
```

clayjohn commented on tutorials/assets_pipeline/retargeting_3d_skeletons.rst:
```diff
@@ -0,0 +1,170 @@
+but the internal transforms are different from the definition.
+This option fixes such models by applying transforms on importer.
+
+.. note:: If the imported scene contains objects other than Skeletons, this option may have a bad effect.
```
```suggestion
.. note:: If the imported scene contains objects other than Skeletons, this option may have a negative effect.
```

clayjohn reviewed (commented):
Looks mostly good! I left some suggestions where the language can be cleaned up a bit.

One thing that I did not comment on in the doc is the use of capitals for important terms. I found that I wasn't sure if things were capitalized because they matched names in the animation system (i.e. they are properties or class names) or because they were something the document was referring to multiple times. 

In general, I think it is more clear to readers to:
1. Link to classes when speaking about a class (like SkeletonProfileHumanoid)
2. use backticks to put things in a code block when you are referring to a Godot property or method

TokageItLab commented:
@mhilbrunner @clayjohn I embedded links for notable classes and used code brackets for properties in the importer.

About some camel words:
- Skeleton: This refers to the Skeleton3D class, but it is always "3D" added, which is annoying and I shortened
- Position track: This is used to refer to the animation track, which is clearly a separate type, but it is defined by enum rather than a separate class
- Bone Rest, Bone Name: These exist as `rest` and `name` properties, but there are no properties called `bone_rest` or `bone_name`

mhilbrunner commented:
At this point, I'm happy to merge as-is and iterate where needed. Thanks for these huge additions!
